test/integ/file_upload.cpp

```diff
@@ -0,0 +1,263 @@
+     This file is part of libhttpserver
+     Copyright (C) 2011-2019 Sebastiano Merlino
+
+     This library is free software; you can redistribute it and/or
+     modify it under the terms of the GNU Lesser General Public
+     License as published by the Free Software Foundation; either
+     version 2.1 of the License, or (at your option) any later version.
+
+     This library is distributed in the hope that it will be useful,
+     but WITHOUT ANY WARRANTY; without even the implied warranty of
+     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+     Lesser General Public License for more details.
+
+     You should have received a copy of the GNU Lesser General Public
+     License along with this library; if not, write to the Free Software
+     Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
+     USA
+*/
+
+#include <curl/curl.h>
+#include <map>
+#include <sstream>
+#include <string>
+
+#include "./httpserver.hpp"
+#include "httpserver/string_utilities.hpp"
+#include "./littletest.hpp"
+
+using std::string;
+using std::map;
+using std::shared_ptr;
+using std::vector;
+using std::stringstream;
+
+using httpserver::http_resource;
+using httpserver::http_request;
+using httpserver::http_response;
+using httpserver::string_response;
+using httpserver::file_response;
+using httpserver::webserver;
+using httpserver::create_webserver;
+
+static const char* TEST_CONTENT_FILENAME = "test_content";
+static const char* TEST_CONTENT_FILEPATH = "./test_content";
+static const char* FILENAME_IN_GET_CONTENT = "filename=\"test_content\"";
+static const char* TEST_CONTENT = "test content of file\n";
+static const char* TEST_KEY = "file";
+static size_t TEST_CONTENT_SIZE = 21;
+
+static CURLcode send_file_to_webserver() {
+    curl_global_init(CURL_GLOBAL_ALL);
+
+    CURL *curl = curl_easy_init();
+
+    curl_mime *form = curl_mime_init(curl);
+    curl_mimepart *field = curl_mime_addpart(form);
+    curl_mime_name(field, TEST_KEY);
+    curl_mime_filedata(field, TEST_CONTENT_FILEPATH);
+
+    CURLcode res;
+    curl_easy_setopt(curl, CURLOPT_URL, "localhost:8080/upload");
+    curl_easy_setopt(curl, CURLOPT_MIMEPOST, form);
+
+    res = curl_easy_perform(curl);
+
+    curl_easy_cleanup(curl);
+    curl_mime_free(form);
+    return res;
+}
+
+class print_file_upload_resource : public http_resource {
+ public:
+     explicit print_file_upload_resource(std::map <std::string, std::string, httpserver::http::arg_comparator>* args,
+                                         std::map<std::string, std::map<std::string, httpserver::http::file_info>>* files,
+                                         stringstream* content) {
+         this->args = args;
+         this->content = content;
+         this->files = files;
+     }
+
+     const shared_ptr<http_response> render_POST(const http_request& req) {
+         (*content) << req.get_content();
+         (*args) = req.get_args();
+         (*files) = req.get_files();
```

Comment:
Can we manage these class attributes as internal values and not use pointers (better with setters/getters, acceptable with a public interface)?

Ideally, from the outside, you can use the getters.